### ISSUE TITLE
Evaluating PoW to avoid a 51% attack

### ISSUE THREAD

**Issue by PaulSnow** [pegnet/pegnet#108]
We have an advantage because the selection of the record that matters is a distributed problem (over all entries submitted).  But any miner with 51% of the hash power still has the chance of selecting the values actually used in a block.  How do we protect ourselves from this?

An analysis of the final 50 OPRs in the current approach has no method to do much but calculate the agreement between the 50 OPRs.  A miner with 26 Entries in the 50 can dictate that result.  So how can we make it harder to get 26 Entries?

Change how we reduce 100 to 200 entries to 50.

This method works like this:
```
Collect the valid OPRs (all references to OPRs past this step assumes the list of valid OPRs)

Calculate the PoW for all the OPRs.

Take the difficulty of the last OPR submitted, and use it to create a salted hash for all OPRs
Sort by Salted Hash.

Then loop through all the OPRs by pairs
 Keep the OPR of the pair that has the highest PoW
    if all that is left + what you are to keep == 50, you are done
    If at the end of the list (without a pair), and we still have more than 50 OPRs, repeat the loop 
         with the OPRs we Kept
```

What this does for any set of valid OPRs over 100 is ensure a party submitting 26 OPRs has a much reduced chance of being in set of 50.  A mining pool submitting multiple entries is likely to compete with themselves prior to the selection of 50.  

To have a good chance to have 26 entries out of 50, 51% is no longer enough.  Many of your entries will end up competing with your own entries, ensuring one or the other no longer counts, no matter how high the hash power is for each.

The only entry with 100% certainty to win and go into the 50 is the highest hash power.  But the second highest hash power might have been paired with the highest and eliminated.  The impact of the algorithm is rather hard for me to calculate.  Someone with some statistics might be able to figure it out.  I need my stats book to do stats.

**Comment by WhoSoup** [comment]
Alright, took a stab at evaluation this and it seems alright. I plotted out the likelihood of getting 26 into the top 50 results using this algorithm:

https://i.imgur.com/W9zOP2z.png

Blue = Your 26 entries are the ones with the highest POW
Yellow = You have a random set of 26 in the top N
Red = Your 26 entries have the worst POW 

Findings:
* If there are more than 54 oprs, it's always better to have higher POW to get your 26 inside the top 50
* With more than 75 oprs, you pretty much need to have the top or near-top POW to get your 26 into the top 50
* If you have the top 26 POW, there's always a chance to perform a 51% attack
* Due to the randomness, likelihood seems to bottom out after 137 OPRs to swing between ~0.2% and ~0.3%

The worst case for the attacker is at more than 137+ OPRs where the likelihood of the highest 26 making it into the top50 is around ~0.25%. I graphed the likelihood of a **sustained attack over X blocks** succeeding: (major gridlines = 1 day, minor gridline = 6 hours)

https://i.imgur.com/KaEHoCh.png

If there are 137+ OPRs and you want to perform a successful 51% attack, there's only 50% likelihood you're going to be successful if you sustain the attack for 288 blocks (2 days). If there's around 100 total oprs, you have a fifty-fifty shot of being successful after ~6 hours.

All in all, it does what it's supposed to: make it more difficult to perform a 51% attack.

**Comment by WhoSoup** [comment]
If we do the algorithm except for the part where we pseudo-shuffle the records, it's a much better outcome, I think. An attacker would have to get the highest difficulty, third highest, fifth highest, etc. in that specific arrangement in order to get their 26 into the top 50. 

The drawback is that you are punishing half of the people with high POW. Thoughts on not pseudo-randomizing the set?

**Comment by PaulBernier** [comment]
Just want to say I am a bit wary about this pseudo random selection of winners. It's could be pretty discouraging for serious miners. Also I think it's risky to change the core algorithm of winner selection for M1. We can definitively keep discussing the pros and cons though or other algos.

**Comment by AroundTheBox** [comment]
Really interesting idea, but here are a few potential issues I see, for what it's worth:

1. Using the last OPR submitted to create salted hashes could cause disagreement to which OPR came last, when many will be submitted at the last second. I think a slightly better approach would be that the OPR with the highest POW value creates the salted hash, although this causes its own problems.

2. An attacker monitoring the OPRs being submitted could potentially manipulate the rankings by attempting to submit the last OPR. They would then be able to calculate the salted hash for all submissions up to that point, contingent on them being successful submitting the last one. During the 10 minute block they could create millions of valid OPRs (although with very low POW) with a very wide spread of salted hashes. They could then take advantage of their information asymmetry and submit only the entries that ensure their actual submissions get favorable pairings. There's no downside to this attack since if they are unsuccessful in submitting the last OPR they haven't lost anything - their odds are the same as they otherwise would have been.

3. It only takes a successful 51% attack for one block to bring down the system, since an attacker could convert from one pegged token to another at extremely favorable rates. By adding randomness you're reducing the probability of someone with 51% of the hashpower getting 25 blocks but increasing the probability of someone with less than 50% getting 25 of the top 50 for any particular block. As an aside - you actually "only" need 25 out of the top 50 to guarantee convergence on your price. If an attacker has 25 fake entries all saying BTC is exactly $100,000 and the other 25 honest entries all have a little variance, saying $10,001 or $9,999, for example, the first entry weeded out is guaranteed to be an honest one, and then every interaction after that shifts to the attacker's side.

Basically I think the original proposal is more elegant and safe.

**Comment by WhoSoup** [comment]
> 1. Using the last OPR submitted to create salted hashes could cause disagreement to which OPR came last, when many will be submitted at the last second. I think a slightly better approach would be that the OPR with the highest POW value creates the salted hash, although this causes its own problems.

Grading happens after the block is finished, meaning there will not be any last second additions. Inside the block itself, the order is not random and the same for everyone. The last OPR submitted will be the same for every miner.

> 2\. An attacker monitoring the OPRs being submitted could potentially manipulate the rankings by attempting to submit the last OPR. They would then be able to calculate the salted hash for all submissions up to that point, contingent on them being successful submitting the last one. During the 10 minute block they could create millions of valid OPRs (although with very low POW) with a very wide spread of salted hashes.

Paul's algorithm is for reducing the top 200 down to the top 50 difficulties, so submitting millions of OPRs would not be any different than submitting your own personal top 200 in that sense. And if you can get enough hashes inside the global top 200 to do what you're talking about, then it's because you have a ton of hash power.

I've done some analysis on the algorithm and entries with higher hashpower are much more likely to survive the pairings, so "getting an even spread" inside the top 200 would not be beneficial, I think. Please feel free to continue pondering this with the highest 200 difficulties limit.


> 3\. It only takes a successful 51% attack for one block to bring down the system, since an attacker could convert from one pegged token to another at extremely favorable rates.

Yeah, that's an unfortunate reality that is true for all proof of work chains. This algorithm wasn't meant to fix the problem but rather increase the cost of doing a 51% attack. Instead of buying enough hash power for just one block, you'd need to sustain the attack for several days.

Anyway, thanks for looking into this algorithm. Have you tried to actually do the math you were talking about in 2\.? I'd be interested to see if it can be done in under a minute.

**Comment by AroundTheBox** [comment]
Thanks for the response. Regarding your question on the attack and computational power, a very low key version that doesn't attempt to manipulate the entire tournament would be as follows:

You have Record A which you'd like to push to the top 50 even though its POW score might not be quite good enough.

Step 1 - Create a valid Record B. It will likely have low POW but doesn't matter.
Step 2 - Calculate salted hash for Record A using Record B's POW.
Step 3 - Generate (but don't submit) a million "strawman" records and calculate the salted hash for each using Record B's POW. Rank them by salted hash and discard all but the 10 closest to Record A's value.
Step 4  - Submit record A and the 10 strawman records. Record A will be paired with the strawman entries and will likely make the top 50.
Step 5 - At the last possible second submit Record B so its POW gets used as the salted hash.

The two flawed assumptions in the plan above are:
1 - It assumes all submitted records would be matched using Paul's tournament algorithm, but it sounds like the proposal was actually to do a straight sort to top 200, tournament to top 50, and then grading algorithm to get to top 10.
2 - It presumes you have a good change of getting the last record in the block if you time it right. I probably have a misunderstanding / lack of understanding for how consensus is achieved that an OPR (or any Factom transaction really) is submitted in time using time delineated blocks.

I guess this is all a moot point if the proposal isn't going forward, but I think it's an interesting idea anyway.

**Comment by WhoSoup** [comment]
> You have Record A which you'd like to push to the top 50 even though its POW score might not be quite good enough.
> 
> Step 1 - Create a valid Record B. It will likely have low POW but doesn't matter.
> Step 2 - Calculate salted hash for Record A using Record B's POW.
> Step 3 - Generate (but don't submit) a million "strawman" records and calculate the salted hash for each using Record B's POW. Rank them by salted hash and discard all but the 10 closest to Record A's value.
> Step 4 - Submit record A and the 10 strawman records. Record A will be paired with the strawman entries and will likely make the top 50.
> Step 5 - At the last possible second submit Record B so its POW gets used as the salted hash.

Hm, I think you have some wrong assumptions on the data the salting produces. For the sake of argument, let's assume we're running the algorithm on all valid OPRs instead of just the top 200 AND that you inspect the process list so you have full knowledge of all OPRs that are about to be submitted.

You have X OPRs and you want to reduce it down to 50. Your goal as attacker is to get 25 of those 50 spots.

Since you know all the OPRs, you make the "last opr" and sort by that salt. What this does is put all entries into a pseudo-random but deterministic order.

In the first pass, you compare records 0 and 1, 2 and 3, 4 and 5, etc, keeping the one with higher difficulty, leaving you with X/2 records. To survive the first pass, you have to match up your OPRs with one of lesser difficulty. To survive the second pass, you have to make sure your OPR also has a higher difficulty than the winner of the adjacent pair. If you're familiar with trees, you can look at this as a binary tree with each node being the child with higher difficulty.

When X = 200, there are two reduction passes.

So essentially you have two requirements:
* Each of your 25 OPRs has to be able to beat 3 OPRs in difficulty (for X=200, they'd have to be inside the range 1-125)
* You need to find those 25 OPRs so that when they are salted, no two oprs are in the same "group" (for X=200, a group is 4 adjacent oprs)

Now this is theoretically possible to find a last OPR with a difficulty that will sort the others into that precise arrangement, but unless you know of a way to break the LXRHash algorithm, it's exceptionally unlikely to do in under a minute.

I hope that makes sense.

**Comment by AroundTheBox** [comment]
Yes it makes sense and I agree with what you said.

My strategy was in the opposite direction to save computation, in that you start with a random end OPR to get the pseudorandom rankings, and then pad the tree as necessary by adding strawman OPRs when needed.

Either way it certainly doesn’t work when you limit to top 200 POW and I’ve taken enough of your time already. Thanks again!

**Comment by Cahl-Dee** [comment]
This has been merged with the grader refactor discussion, see #259
